libqrexec/exec.c

```diff
@@ -105,4 +139,240 @@ void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
+    const char *const username;
+    /* Command line.  Never NULL.  NUL-terminated. Does not include "QUBESRPC ". */
+    char *const command;
+    /* Service descriptor.  Identical to `command` unless the "nogui:" prefix is present, in which case it points
+     * after the colon.  Always points to the start of the service name. */
+    const char *service_descriptor;
+    /* Remote domain.  NUL-terminated.  Always equal to `service_descriptor + service_descriptor_length + 1`. */
+    const char *const remote_domain;
+    /* Size of service_descriptor (the service name + argument).  Guaranteed to be <= MAX_SERVICE_NAME_LEN. */
+    size_t const service_descriptor_length;
+};
+
+static int execute_parsed_qubes_rpc_command(const struct qrexec_parsed_command *const command, int *const pid, int *const stdin_fd, int *const stdout_fd, int *const stderr_fd, struct buffer *stdin_buffer);
+
+static const char *skip_nogui(const char *cmdline) {
+    return strncmp(cmdline, NOGUI_CMD_PREFIX, NOGUI_CMD_PREFIX_LEN) ? cmdline : cmdline + NOGUI_CMD_PREFIX_LEN;
+}
+
+int execute_qubes_rpc_command(char *cmdline, int *pid, int *stdin_fd,
+        int *stdout_fd, int *stderr_fd, bool strip_username, struct buffer *stdin_buffer) {
+    const char *service_descriptor, *remote_domain;
+    char *realcmd;
+    size_t service_descriptor_length;
+    {
+        char *end_user;
+        if (strip_username) {
+            end_user = strchr(cmdline, ':');
+            if (!end_user) {
+                fputs("Bad command from dom0: no colon\n", stderr);
+                abort();
+            }
+            *end_user++ = '\0';
+        } else {
+            end_user = cmdline;
+        }
+        if (strncmp(end_user, RPC_REQUEST_COMMAND " ", RPC_REQUEST_COMMAND_LEN + 1) != 0) {
+            /* Legacy qrexec behavior: spawn shell directly. */
+            return do_fork_exec(strip_username ? cmdline : NULL, end_user, pid, stdin_fd, stdout_fd, stderr_fd);
+        } else {
+            realcmd = end_user + RPC_REQUEST_COMMAND_LEN + 1;
+            service_descriptor = skip_nogui(realcmd);
+        }
+    }
+    remote_domain = strchr(service_descriptor, ' ');
+    if (!remote_domain) {
+        fputs("Bad command from dom0: no remote domain\n", stderr);
+        abort();
+    }
+    remote_domain++; // skip the space character
+    if (strlen(remote_domain) > 64) {
+        fputs("Bad command from dom0: remote domain too long\n", stderr);
+        abort();
+    }
+    service_descriptor_length = (size_t)(remote_domain - service_descriptor);
+    /* Check that the path is of a valid length */
+    if (service_descriptor_length > MAX_SERVICE_NAME_LEN) {
+        fprintf(stderr, "Bad command from dom0: absurdly long command (length %zu)\n", service_descriptor_length);
+        abort();
+    }
+    const struct qrexec_parsed_command command = {
+       .username = strip_username ? NULL : cmdline,
```

Comment:
This is swapped. If we are stripping the username then we know it.

Reply:
Fixed.